### ISSUE TITLE
[gcov] partial processing

### ISSUE THREAD

**Issue by rastislavszabo** [codecov/support#193]
It seems that codecov now recognizes partial hits in the code, which is a nice feature. However, I think that the effect of partial hits on overall coverage, where they count as misses is not very well defined. 

E.g.: https://codecov.io/gh/sysrepo/sysrepo/src/e25f41b0dfef67feb46ed686a72bb38615f6e661/src/client_library.c

447 hits
197 partial hits
107 misses
745 total statements

Shows coverage 60%, which is 447 out of 745. That means that partial hits do not count at all.

Also, `if` statements like this one are for some reason incorrectly considered as partials:
`if (NULL != tmp) {`

I think that the best solution would be to enable users to configure how partials should be counted:
a) count them as hits
b) count them as misses
c) count how much of the code is covered in partial blocks (e.g. calling a macro that contains 10 statements with 7 of them covered should count as 0.7 of a hit)

Because of the recent change, the coverage of our project dropped from 83% to 57% during the last day. When such a change happens from one day to another, it's quite tough to report our progress of covering the code with unit tests.

**Comment by stevepeak** [comment]
Hey so we did start recognizing the branch arguments provided in `gcov` as of our v4 release lately.

First lets load up the report [here](https://codecov.s3.amazonaws.com/v4/raw/2016-05-10/46DBA4BC63527E8DB7BECD7F3FD9CEED/e25f41b0dfef67feb46ed686a72bb38615f6e661/829fa670-49f3-4887-ab7e-3a29a72706b8.txt) (which can be found in codecov ui under **builds**)

Once we have the report lets search for `CHECK_NULL_ARG2(conn_ctx, socket_path);`. You'll see the results is:

```54:   73:    CHECK_NULL_ARG2(conn_ctx, socket_path);
branch  0 taken 0% (fallthrough)
branch  1 taken 100%
branch  2 never executed
....
branch 42 never executed
call   43 never executed
```

According to the there are a number of branches not taken on this line. I would like to review this line with you to learn more about understanding the gcov data.
 
-----

Next line is https://codecov.io/gh/sysrepo/sysrepo/src/e25f41b0dfef67feb46ed686a72bb38615f6e661/src/client_library.c#L159

<img width="642" alt="screen shot 2016-05-16 at 9 35 58 am" src="https://cloud.githubusercontent.com/assets/2041757/15290955/a9fac4a4-1b49-11e6-8e18-acccfb545564.png">

The gcov report expresses:

```
        9:  159:    if (NULL != module_name) {
branch  0 taken 56% (fallthrough)
branch  1 taken 44%
branch  2 never executed
branch  3 never executed
```
> The branch at 2 and 3 are stated as never taken. However this is very strange because the other branches presumable claim 100% of the calls.

-----

I would like to discuss with you about how we can best "fix" these two issues.

**Comment by stevepeak** [comment]
New `codecov.yml` additions coming soon:

```yaml
# default setup
coverage:
  parsers:
    gcov:
      branch_detection:
        macro: no
        loop: yes
        conditional: yes
```

**Comment by stevepeak** [comment]
@rastislavszabo this has been deployed. The above configuration is the default now for gcov. Let me know if the implentation works for you.

**Comment by rastislavszabo** [comment]
@stevepeak It seems to be working fine! Thanks!
Only a very minor issue that I have encountered. Somewhere in your documentation I've red that I can also use ".codecov.yml" (dot as the first character) file name, but it seems to be not working.

**Comment by Flamefire** [comment]
I'm having a similar problem:
In https://codecov.io/gh/Return-To-The-Roots/s25client/src/806bc337994cc583cd031d97e4d4f3c06ed4addd/src/world/BQCalculator.h it shows a partial coverage for `if(bq == TerrainBQ::CASTLE)` (comparison involving a scoped enum) GCov seems to report 5(!!!) branches for this, instead of the expected 2. I don't see anything I can do about this but completely disable partial coverage (I guess this is possible with the above config options)

**Comment by stevepeak** [comment]
@Flamefire We have implemented the `gcov` special detection via the yaml. But we do not detect this special enum inside a condition. 

How would that pattern look? We use regexp to determin the what type of line it is. Would we use something like: `(\w+\:\:[A-Z]+)`. If this pattern is found then we should limit the branch detection to only 2 branches.

If detected then which branches should we ignore in the branch data below:

```
  112854:   49:        if(bq == TerrainBQ::CASTLE)
branch  0 never executed
branch  1 never executed
branch  2 never executed
branch  3 never executed
branch  4 taken 55% (fallthrough)
branch  5 taken 45%
```

**Comment by Flamefire** [comment]
I got to doublecheck this, as this might be caused by the different template instantiations. This would explain, why I can't reproduce this and the report would be correct.
About branches: Basically an `if(foo <cmp-op> bar)` should only generate 2 branches: fallthrough and the other one (4 and 5 in this case). But in this case it might be 3 pairs of 2 branches.

**Comment by rbaumbach** [comment]
@stevepeak I'm late to the convo, but are these new additions documented somewhere?  I didn't see them under the interactive "Yaml / Settings" for a repo.  I'm not too familiar with most of the yaml options at this point.

**Comment by stevepeak** [comment]
Hey @rbaumbach essentially the [comment above](https://github.com/codecov/support/issues/193#issuecomment-219440284) is what I've added to Codecov.

I'm happy to help you out. Can you explain a little more on what your are trying to achieve? Feel free to chat with me in Intercom on Codecov. Thanks!

**Comment by stevepeak** [comment]
@Flamefire this is such a tough one to solve. How can we know what branches to use, and which to ignore?

Take this for example:

```
  1128470:   62:    if(flag_hits)
branch  0 taken 0% (fallthrough)
branch  1 taken 100%
branch  2 taken 0% (fallthrough)
branch  3 taken 100%
branch  4 taken 0% (fallthrough)
branch  5 taken 100%
branch  6 taken 0% (fallthrough)
branch  7 taken 100%
branch  8 taken 2% (fallthrough)
branch  9 taken 98%
```

> This would result in `6/10` partial hit. But when [looking at the code](https://codecov.io/gh/Return-To-The-Roots/s25client/src/master/src/world/BQCalculator.h#L62) its clearly a hit.

What about the concept of "best two branches"? So in this case it would use branch 8+9. Resulting in a `2/2` hit.

-----

```
  112854:   49:        if(bq == TerrainBQ::CASTLE)
branch  0 never executed
branch  1 never executed
branch  2 never executed
branch  3 never executed
branch  4 taken 55% (fallthrough)
branch  5 taken 45%
```
> Would use branch 4+5` resulting in `2/2` hit.

Thoughts?

**Comment by stevepeak** [comment]
What would this be....?

```
   28957:  118:    } else if(curBQ == BQ_MINE && world.GetNeighbourNode(pt, 4).altitude > curAltitude + 3)
branch  0 taken 1% (fallthrough)
branch  1 taken 99%
call    2 returned 100%
call    3 returned 100%
branch  4 taken 100% (fallthrough)
branch  5 taken 0% (throw)
branch  6 taken 0% (fallthrough)
branch  7 taken 100%
branch  8 taken 1% (fallthrough)
branch  9 taken 99%
branch 10 taken 0% (fallthrough)
branch 11 taken 100%
branch 12 never executed
branch 13 never executed
branch 14 taken 0% (fallthrough)
branch 15 taken 100%
call   16 never executed
call   17 never executed
branch 18 never executed
branch 19 never executed
branch 20 never executed
branch 21 never executed
branch 22 taken 0% (fallthrough)
branch 23 taken 100%
branch 24 taken 0% (fallthrough)
branch 25 taken 100%
branch 26 never executed
branch 27 never executed
branch 28 taken 0% (fallthrough)
branch 29 taken 100%
call   30 never executed
call   31 never executed
branch 32 never executed
branch 33 never executed
branch 34 never executed
branch 35 never executed
branch 36 taken 0% (fallthrough)
branch 37 taken 100%
branch 38 taken 0% (fallthrough)
branch 39 taken 100%
branch 40 never executed
branch 41 never executed
branch 42 never executed
branch 43 never executed
call   44 never executed
call   45 never executed
branch 46 never executed
branch 47 never executed
branch 48 never executed
branch 49 never executed
branch 50 never executed
branch 51 never executed
branch 52 never executed
branch 53 never executed
branch 54 never executed
branch 55 never executed
branch 56 taken 66% (fallthrough)
branch 57 taken 34%
call   58 returned 100%
call   59 returned 100%
branch 60 taken 100% (fallthrough)
branch 61 taken 0% (throw)
branch 62 taken 11% (fallthrough)
branch 63 taken 89%
branch 64 taken 66% (fallthrough)
branch 65 taken 34%
branch 66 taken 7% (fallthrough)
branch 67 taken 93%
branch 68 never executed
branch 69 never executed
```

**Comment by Flamefire** [comment]
I think the current implementation is ok. IMO the different branches are caused by template instantiations. So you have actually 3 or 5(?) code paths assigned to this line (each a template instantiation). And in only 1 both conditions are evaluated at least once. Similar for the last example. Although there it also counts inlined code from `GetNeighbourNode` as it seems.
I can't see a better solution for now. Probably an option could be possible, to collapse all template instantiations. But I think this is not possible with only the coverage data.

**Comment by Flamefire** [comment]
Hi @stevepeak . As it seems this is very complex (and the above being a template-issue), I'd like to start with simpler cases. I got a line, that is detected as partial because it calls some functions. So the `if`-conditions are covered, but somehow the functions are not completely covered. Probably partially system/boost accessors with error throwing (which should be ignored for coverage).
I can't make up anything from the gcov files and also opened a question on SO for possible help on compiler options: http://stackoverflow.com/questions/38478916/unexpected-branch-count-in-coverage-data

The line is `if(nodeBQ == BQ_NOTHING || GetNode(pt).owner != player + 1 || !IsPlayerTerritory(pt))` and I get 24 branches and calls! Please check the repo or the SO question for details or tell me if you need further information. What I would want from this: 6 branches, each covered or not (in this case they are) and a full hit for that.

There is not even a good pattern :( (->My guesses for the valid branches)
```
branch  0 taken 95% (fallthrough) --> VALID
branch  1 taken 5%                --> VALID
call    2 returned 100%
call    3 returned 100%
branch  4 taken 100% (fallthrough)
branch  5 taken 0% (throw)
branch  6 taken 85% (fallthrough)  --> VALID
branch  7 taken 15%                --> VALID
call    8 returned 100%
call    9 returned 100%
branch 10 taken 100% (fallthrough)
branch 11 taken 0% (throw)
branch 12 taken 17% (fallthrough)  --> VALID
branch 13 taken 83%                --> VALID
branch 14 taken 80% (fallthrough)
branch 15 taken 20%
branch 16 taken 95% (fallthrough)
branch 17 taken 5%
branch 18 taken 33% (fallthrough)
branch 19 taken 67%
branch 20 never executed
branch 21 never executed
branch 22 never executed
branch 23 never executed
call   24 never executed
```

**Comment by stevepeak** [comment]
Hey @Flamefire this leaves me puzzled on how to accurately predict/analyze which branches are valid. I see no correlation in this data that points to a means of filtering the invalid branches.

Maybe we should jump on a chat/call this week to hash this out? I've invited you to our Slack room where you can find me. Thanks!

**Comment by Flamefire** [comment]
Yes I do see the problem, but I don't have any solution which is why I opened the SO question hoping for some input from compiler-relatated guys. My guesses above are based on the 2nd example where the if only checks the conditions but does not call any function.

I may be able to join Slack on Tuesday at about 7PM (GMT+1) but I'm not sure, if I can be of much help

**Comment by Flamefire** [comment]
@stevepeak I just found that clang produces much better results. Looking at the gcov files I see exactly the branches I'd expect. Let's see if I can get travis to run those

**Comment by stevepeak** [comment]
```
-l, --long-file-names
              For coverage output of files included from the main source file, add the main file name followed by ## as a prefix to the output file names.
              This can be combined with the --preserve-paths option to use complete paths for both the main file and the included file.
```

This options is not required for Codecov to properly map the files back. I actually have not tried this flag out before.

**Comment by Flamefire** [comment]
Ok, clang seems to help. But there still are problems: https://codecov.io/gh/Return-To-The-Roots/s25client/compare/14796d1cb7423a87f9698001db650fa6262ae4ee...bc1c6d21d10b9db2892f8e7df69d228d7205c2d6/src/src/ingameWindows/iwPostWindow.cpp (from https://github.com/Return-To-The-Roots/s25client/pull/507) -> A "covered" line between uncovered ones. I also see some uncovered lines in headers that are called from the tests. I suspect inlining to be the problem, but need further checks.

**Comment by rianquinn** [comment]
@stevepeak I'm seeing the same issue here with GCC. As an example, the following code shows partials all over the place:

![](https://s29.postimg.org/snubz4cfr/Screen_Shot_2016_12_08_at_3_15_54_PM.png)

I see it in other places too, but this is a good example as there are only two paths it can take, and clearly both are taken as you can see the taken lines on both paths, and it's reporting 4 paths, which also doesn't make much sense to me. I'll try clang and report back.

**Comment by rianquinn** [comment]
@stevepeak Getting Clang to work with Travis is a real pain, and then in addition, the Clang version has to match gcov, otherwise it gets really mad. For now, the best solution I had was to turn off the branch detection in the YML config.

**Comment by jasonmp85** [comment]
Is the `gcov` `yml` configuration documented anywhere?

**Comment by stevepeak** [comment]
@jasonmp85 I'm not aware of a gcov yml. Are you looking for [codecov.yml](https://docs.codecov.io/docs/codecov-yaml)?

We will be introducing a new configuration to count partials as `hit`, `miss` or `ignore` soon.

**Comment by jasonmp85** [comment]
> I'm not aware of a gcov yml. Are you looking for codecov.yml?

No, I meant the configuration options hierarchically located beneath the `gcov` node inside of `codecov.yml` (such as the changes where you referenced this issue earlier in the thread). `branch_detection`, etc. It's clear there are some configuration options that can appear nested under a `gcov` node but I didn't remember seeing them documented when I wrote that comment.

**Comment by gsauthof** [comment]
With a C++ project of mine I also observe this issue. Similar to the example @rianquinn posted, the partials are all over the place:

https://codecov.io/gh/gsauthof/imapdl/src/46cc2ce1896ad64180880d2e735dabda5850eee7/maildir/maildir.cc

The coverage data comes from a build job that compiles with GCC and uses lcov to create a .info file which is explicitly uploaded to codecov.io. Lcov is configured to include branch coverage data (lcov excludes it, by default).

The issue with C++ is that basically at each line where an exception induces scope exit may happen this line is recorded as 2-way branch by GCC - and thus displayed as partial in codecov.

When generating a html report with lcov's genhtml the same lines that codecov displays as partials are displayed as 2-way branches. In contrast to codecov, those lines are still included in the computation of the relative coverage values. For example, for the referenced example file the lcov report shows a function coverage of 94 % while codecov shows 50 %.

What I want is this: configure codecov to behave similar to lcov's genhtml, i.e. tell codecov to count 'partial' lines also as covered but still display partials as such in the source code view. Is this possible?

Using a `codcov.yml` like this:

    parsers:
      gcov:
        branch_detection:
          conditional: no
          loop: no
          method: no
          macro: no

doesn't make a difference (results for the corresponding commit:  7de05335391826564e173a8328ddb9e4745d8e2d). The example file is still at 50 % and not at 94 %, as expected.

I didn't find documentation for the above keys - so I'm not really how they are supposed to work.

See also e.g.
http://stackoverflow.com/questions/42003783/lcov-gcov-branch-coverage-with-c-producing-branches-all-over-the-place
